Tasks/Gulp/gulptask.ts

```diff
@@ -2,118 +2,154 @@
+			if (isCodeCoverageEnabled) {
+				return npm.exec().then(function () {
+					return istanbul.exec().then(function (code) {
+						publishCodeCoverage(summaryFile, reportDirectory);
+						console.log(tl.loc('GulpReturnCode', code));
+					}, function (err) {
+						publishCodeCoverage(summaryFile, reportDirectory);
+						tl.debug('taskRunner fail');
+						throw new Error(tl.loc('IstanbulFailed', err.message));
+					})
+				}, function (err) {
+					tl.debug('taskRunner fail');
+					throw new Error(tl.loc('NpmFailed', err.message));
+				})
+			}
+			else {
+				console.log(tl.loc('GulpReturnCode', code));
+			}
+		}, function (err) {
+			publishTestResults(publishJUnitResults, testResultsFiles);
 			tl.debug('taskRunner fail');
-			tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+			throw new Error(tl.loc('GulpFailed', err.message));
 		})
-	} else {
-		tl.setResult(tl.TaskResult.Succeeded, tl.loc('GulpReturnCode', code));
 	}
-}).fail(function (err) {
-	publishTestResults(publishJUnitResults, testResultsFiles);
-	tl.debug('taskRunner fail');
-	tl.setResult(tl.TaskResult.Failed, tl.loc('GulpFailed', err.message));
-})
+}
 
 function publishTestResults(publishJUnitResults, testResultsFiles: string) {
-    if (publishJUnitResults) {
-        //check for pattern in testResultsFiles
-        if (testResultsFiles.indexOf('*') >= 0 || testResultsFiles.indexOf('?') >= 0) {
-            tl.debug('Pattern found in testResultsFiles parameter');
-            var buildFolder = tl.getVariable('System.DefaultWorkingDirectory');
-            var allFiles = tl.find(buildFolder);
-            var matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, { matchBase: true });
-        }
-        else {
-            tl.debug('No pattern found in testResultsFiles parameter');
-            var matchingTestResultsFiles = [testResultsFiles];
-        }
-        if (!matchingTestResultsFiles || matchingTestResultsFiles.length == 0) {
-            tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
-            return 0;
-        }
-        var tp = new tl.TestPublisher("JUnit");
+	if (publishJUnitResults) {
+		//check for pattern in testResultsFiles
+		if (testResultsFiles.indexOf('*') >= 0 || testResultsFiles.indexOf('?') >= 0) {
+			tl.debug('Pattern found in testResultsFiles parameter');
+			var buildFolder = tl.getVariable('System.DefaultWorkingDirectory');
+			var allFiles = tl.find(buildFolder);
+			var matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, "", { matchBase: true });
+		}
+		else {
+			tl.debug('No pattern found in testResultsFiles parameter');
+			var matchingTestResultsFiles = [testResultsFiles];
+		}
+		if (!matchingTestResultsFiles || matchingTestResultsFiles.length == 0) {
+			tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
```

Comment:
Consider using tl.findMatch directly. This way both the pattern matches will work the same way